crates/config/src/endpoints.rs

```diff
@@ -37,13 +49,99 @@ impl RpcEndpoints {
 }
 
 impl Deref for RpcEndpoints {
-    type Target = BTreeMap<String, RpcEndpoint>;
+    type Target = BTreeMap<String, RpcEndpointType>;
 
     fn deref(&self) -> &Self::Target {
         &self.endpoints
     }
 }
 
+/// RPC endpoint wrapper type
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum RpcEndpointType {
+    String(RpcEndpoint),
+    Config(RpcEndpointConfig),
+}
+
+impl RpcEndpointType {
+    /// Returns the string variant
+    pub fn as_endpoint_string(&self) -> Option<&RpcEndpoint> {
+        match self {
+            RpcEndpointType::String(url) => Some(url),
+            RpcEndpointType::Config(_) => None,
+        }
+    }
+
+    /// Returns the config variant
+    pub fn as_endpoint_config(&self) -> Option<&RpcEndpointConfig> {
+        match self {
+            RpcEndpointType::Config(config) => Some(&config),
+            RpcEndpointType::String(_) => None,
+        }
+    }
+
+    /// Returns the url or config this type holds
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the type holds a reference to an env var and the env var is not set
+    pub fn resolve(self) -> Result<String, UnresolvedEnvVarError> {
+        match self {
+            RpcEndpointType::String(url) => url.resolve(),
+            RpcEndpointType::Config(config) => config.endpoint.resolve(),
+        }
+    }
+}
+
+impl fmt::Display for RpcEndpointType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RpcEndpointType::String(url) => url.fmt(f),
+            RpcEndpointType::Config(config) => config.fmt(f),
+        }
+    }
+}
+
+impl TryFrom<RpcEndpointType> for String {
+    type Error = UnresolvedEnvVarError;
+
+    fn try_from(value: RpcEndpointType) -> Result<Self, Self::Error> {
+        match value {
+            RpcEndpointType::String(url) => url.resolve(),
+            RpcEndpointType::Config(config) => config.endpoint.resolve(),
+        }
+    }
+}
+
+impl Serialize for RpcEndpointType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            RpcEndpointType::String(url) => url.serialize(serializer),
+            RpcEndpointType::Config(config) => config.serialize(serializer),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for RpcEndpointType {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut value = serde_json::Value::deserialize(deserializer)?;
+
+        if let Some(_obj) = value.as_object_mut() {
+            let config = RpcEndpointConfig::deserialize(value).map_err(serde::de::Error::custom)?;
+            return Ok(RpcEndpointType::Config(config));
+        } else {
+            let endpoint = RpcEndpoint::deserialize(value).map_err(serde::de::Error::custom)?;
+            return Ok(RpcEndpointType::String(endpoint));
```

Comment:
here we can use the enum type you added above this way we're backwards compatible